### PR TITLE
Adds rerouting based on last visited or first page

### PR DESCRIPTION
**Pull request by tuchfarber**
...and all dependancies for that including:
- Header now gathers user information (username, last visited pages)
- Added JournalRerouter which redirects based on logged in andlast visited page
- Add user info API fetching

Also:
- refactored constructors to componentDidMount
- refactored journal state to a flatter structure
- add error handling to fetch actions
- refactor JournalApiService to handle all journals APIs, not just pages
- removed unnecessary exports from actions

### CODE REVIEW

**Comment by tuchfarber** [comment]
Going to refactor to handle protected routing similar to: https://github.com/edx/edx-portal/pull/10/files

File: src/components/JournalRerouter/index.jsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  componentDidMount(){
+    this.props.getJournal(this.props.match.params.journalId);
+  }
+  render(){
+    if(this.props.userInfoFinishedFetching && this.props.journalFinishedFetching){
+      // Only run if both the user info and journal info API calls have finished
+      if(this.props.loggedIn) {
+        if(this.props.lastVisitedPage){
+          // Send user to the last page they visited if available
+          let lastVisitedPageUrl = `/${this.props.match.params.journalId}/pages/${this.props.lastVisitedPage}`;
```
Comment (bill-filler) [inline src/components/JournalRerouter/index.jsx]: where do we set the lastVisitedPage?

Comment (tuchfarber) [reply src/components/JournalRerouter/index.jsx]: https://github.com/edx/journals-frontend/pull/3/files#diff-d9c51a596f54d7ad40565b79574689f8R15
`lastVisitedPage: state.userInfo.visitedPages.length != 0 ? state.userInfo.visitedPages[0].page: null`

It's just taking the PageVisit items returned and grabbing the first one.

Comment (bill-filler) [reply src/components/JournalRerouter/index.jsx]: @tuchfarber 
right but don't think we are setting the PageVisits for a given user anywhere. This needs to be an explicit post to the pagevisit endpoint by the client the way it's currently implemented. See https://github.com/edx/journals/blob/master/journals/apps/api/v1/tests/test_views.py#L28

Comment (bill-filler) [reply src/components/JournalRerouter/index.jsx]: ack

**Comment by tuchfarber** [comment]
Refactored a bit, so this contains some changes that are tangentially related to the rerouting. 
@georgebabey I've implented a PrivateRoute similar to the edx-portal one and will replace once we can create a generic one in the paradigm project.

**Pull request merged**